Basic Extension and Event Support

jonmatterson opened the pull request:
This pull request adds support for two new libraries:

* Events: A library that supports listeners that triggers events and provides callbacks that can listen for an event or a namespace of events.
* Extensions: A library that looks at a config file to load in a set of extensions, each of which are loaded by executing the extension's `init.php` file.

Together, these two libraries make it possible to write extensions installed with only the following three steps:

1. Copy the extension's source files into `application/extensions/EXTENSION_NAME`
2. Add EXTENSION_NAME to `application/config/extensions.php`
3. Update the database as necessary

In the future, it might be nice to remove step (3), but this pull request is just the foundation of it.

Once this occurs, extensions can run that simply hook into the event library, with events supported in this PR to allow for changing data before CRUD operations are performed, as well as to mutate view data before a view is loaded and to modify the output of the view before it's passed back from `Location::view()`. It is likely we will want to add a few more of these events before we release 2.6, likely with events around other `Location` methods (and other places?).

I've included some comments to explain what's going on, but better than that, here are some simple examples...

## Extension Example Code

This example adds two extensions, `jquery` for a little `generator` helper that's exposed via access to the extension container and `bfms_character` which modifies the character page, join form, and edit form to add a new field and display content driven from an external API.

#### `application/config/extensions.php`

```php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once MODPATH.'core/config/nova_extensions.php';

$config['extensions']['enabled'][] = 'jquery';
$config['extensions']['enabled'][] = 'bfms_character';
```

#### `application/extensions/jquery/init.php`

```php
<?php 
require_once dirname(__FILE__).'/libraries/jquery_generator.php';

$this->attach('generator', new jquery_generator());
```

#### `application/extensions/jquery/libraries/jquery_generator.php`

```php
<?php 
class jquery_generator
{
  public function select($selector)
  {
    return new jquery_generator_chain($selector);
  }
}

class jquery_generator_chain
{
  public $selector;
  public $chain = [];
  
  public function __construct($selector)
  {
    $this->selector = $selector;
  }
  
  public function __call($method, $args = [])
  {
    if(in_array($method, ['first','last','before','after','append','prepend','html','text'])){
      $this->chain[] = [
        'type' => 'method',
        'name' => $method, 
        'args' => $args
      ];
    }else{
      show_error('Unsupported jquery_generator_chain method: '.$method);
    }
    return $this;
  }
  
  public function __toString()
  {
    return '<script type="text/javascript">'.implode('.', array_merge([
      '$('.json_encode($this->selector).')'
    ], array_map(function($entry){
      switch($entry['type']){
        case 'method':
          return $entry['name'].'('.implode(',', array_map(function($arg){
            return json_encode($arg);
          }, $entry['args'])).')';
      }
    }, $this->chain))).'</script>';
  }
}
```

#### `application/extensions/bfms_character/init.php`

```php
<?php

require_once dirname(__FILE__).'/helpers/bfms_character_api_helper.php';

$this->require_extension('jquery');

$this->event->listen(['db', 'insert', 'prepare', 'characters', 'main', 'join'], function($event){
  if(empty($event['data']['bfms_character_url']) && $this->input->post('bfms_character_url', true))
    $event['data']['bfms_character_url'] = $this->input->post('bfms_character_url', true);
});

$this->event->listen(['location', 'view', 'data', 'main', 'main_join_2'], function($event){
  $event['data']['label']['bfms'] = 'Bravo Fleet Management System';
  $event['data']['label']['bfms_instructions'] = 'If you provide the URL to your character within BFMS, the data from BFMS will be displayed so that you do not have to duplicate it.';
  $event['data']['label']['bfms_character_url'] = 'Character URL';
  $event['data']['inputs']['bfms_character_url'] = array(
    'name' => 'bfms_character_url',
    'id' => 'bfms_character_url'
  );
});

$this->event->listen(['location', 'view', 'output', 'main', 'main_join_2'], function($event){
  $event['output'] .= $this->extension['jquery']['generator']
                  ->select('#three')
                  ->prepend($this->extension['bfms_character']->view('characters_bio', $this->skin, 'admin', $event['data']));
});

$this->event->listen(['location', 'view', 'data', 'admin', 'characters_create'], function($event){
  $event['data']['label']['bfms'] = 'Bravo Fleet Management System';
  $event['data']['label']['bfms_instructions'] = 'If you provide the URL to your character within BFMS, the data from BFMS will be displayed so that you do not have to duplicate it.';
  $event['data']['label']['bfms_character_url'] = 'Character URL';
  $event['data']['inputs']['bfms_character_url'] = array(
    'name' => 'bfms_character_url',
    'id' => 'bfms_character_url'
  );
});

$this->event->listen(['location', 'view', 'output', 'admin', 'characters_create'], function($event){
  $event['output'] .= $this->extension['jquery']['generator']
                  ->select('#two')
                  ->prepend($this->extension['bfms_character']->view('characters_bio', $this->skin, 'admin', $event['data']));
});

$this->event->listen(['location', 'view', 'data', 'admin', 'characters_bio'], function($event){
  $id = $this->uri->segment(3);
  $char = $this->char->get_character($id);
  $event['data']['label']['bfms'] = 'Bravo Fleet Management System';
  $event['data']['label']['bfms_instructions'] = 'If you provide the URL to your character within BFMS, the data from BFMS will be displayed so that you do not have to duplicate it.';
  $event['data']['label']['bfms_character_url'] = 'Character URL';
  $event['data']['inputs']['bfms_character_url'] = array(
    'name' => 'bfms_character_url',
    'id' => 'bfms_character_url',
    'value' => $char->bfms_character_url
  );
});

$this->event->listen(['location', 'view', 'output', 'admin', 'characters_bio'], function($event){
  $event['output'] .= $this->extension['jquery']['generator']
                  ->select('#two')
                  ->prepend($this->extension['bfms_character']->view('characters_bio', $this->skin, 'admin', $event['data']));
});

$this->event->listen(['location', 'view', 'data', 'main', 'personnel_character'], function($event){
  $id = $this->uri->segment(3);
  $char = $this->char->get_character($id);
  $content = bfms_character__api__get_character($char);
  $event['data']['bfms_character'] = json_decode($content, true);
});

$this->event->listen(['location', 'view', 'output', 'main', 'personnel_character'], function($event){
  $event['output'] .= $this->extension['jquery']['generator']
                  ->select('#tabs')
                  ->append($this->extension['bfms_character']->view('personnel_character', $this->skin, 'main', $event['data']));
});
```

#### `application/extensions/bfms_character/helpers/bfms_character_api_helper.php`

```php
<?php 

if(!function_exists('bfms_character__api__get_character'))
{
  function bfms_character__api__get_character($char)
  {
    $apiUrl = preg_replace('/character\/(\d+)/i', 'wp-json/wp/v2/character/$1', $char->bfms_character_url);
    $ch = curl_init();
    curl_setopt_array($ch, array(
      CURLOPT_RETURNTRANSFER => true,
      CURLOPT_HEADER         => false,
      CURLOPT_USERAGENT      => "nova",
    ));
    curl_setopt($ch, CURLOPT_URL, $apiUrl);
    $content = curl_exec($ch);
    curl_close($ch);
    return $content;
  }
}
```

#### `application/extensions/bfms_character/views/admin/pages/characters_bio.php`

```php
<?php echo text_output($label['bfms'], 'h3', 'page-subhead'); ?>

<div class="indent-left" id="bfms_character_fields">
 
 <?php echo text_output($label['bfms_instructions'], 'p'); ?>
 
 <p>
   <kbd><?php echo $label['bfms_character_url'] ?></kbd>
   <?php echo form_input($inputs['bfms_character_url']) ?>
 </p>
 
 <script type="text/javascript">
 $("#<?php echo $inputs['bfms_character_url']['id'] ?>").change(function(){
  var method;
  if($(this).val().trim().length > 0){
    method = 'hide';
  }else{
    method = 'show';
  }
  $('textarea#history').parent()[method]();
  $('textarea#service_record').parent()[method]();
 }).change();
 </script>
</div><br/>
```

#### `application/extensions/bfms_character/views/main/pages/personnel_character.php`

```php
<?php if($bfms_character): ?>
  
<div id="bfms_history" class="ui-tabs-panel ui-widget-content ui-corner-bottom ui-tabs-hide" style="padding-top:18px;">
  
  <?php echo $bfms_character['bio']['history'];?>
  
</div>
  
<div id="bfms_service_record" class="ui-tabs-panel ui-widget-content ui-corner-bottom ui-tabs-hide">
  
  <table class="table100 zebra">
  <?php foreach($bfms_character['bio']['service_record'] as $item): ?>
    
    <tr>
      <td><?php echo $item['year_start']; ?></td>
      <td><?php echo $item['year_end']; ?></td>
      <td style="text-align:center;line-height:1;">
        <?php if(!empty($item['rank_image_url'])): ?>
        <img src="<?php echo $item['rank_image_url']; ?>" style="margin-top:2px;">
        <br>
        <small style="font-size:11px"><?php echo $item['rank_name']; ?></small>
        <?php else: ?>
        <?php echo $item['rank_name']; ?>
        <?php endif; ?>
      </td>
      <td><?php echo $item['position']; ?><br><em style="font-size:12px"><?php echo $item['assignment']; ?></em></td>
    </tr>
    
  <?php endforeach; ?>
  </table>
  
</div>


<script type="text/javascript">
$('#tabs > ul > li:contains("History")').hide();
$('#tabs > ul').append('<li><a href="#bfms_history"><span>History</span></a></li>');
$('#tabs > ul').append('<li><a href="#bfms_service_record"><span>Service Record</span></a></li>');
</script>
  
<?php endif; ?>
```

## Extension Example Outcome 

The outcome of this is that without requiring the Nova user to make any changes to their Nova controllers or views, the extension is able to modify the tabs on the character page such as:

<img width="1401" alt="screen shot 2018-08-27 at 12 14 02 am" src="https://user-images.githubusercontent.com/2497540/44683311-d37e7b80-a9fa-11e8-8906-e112ce0ca238.png">

And new fields in the edit and join screens (including saving the new fields) such as:

<img width="1376" alt="screen shot 2018-08-27 at 1 13 04 pm" src="https://user-images.githubusercontent.com/2497540/44683353-fa3cb200-a9fa-11e8-96f9-fbe338b2d349.png">

Sure, there's some trickery in how you write the views, inject content into the output string, etc., but that's kind of just a consequence of the lack of modularity. It's still a very powerful way to get around requiring the user to actually copy and paste code, update a skin, etc., to add a feature.

jonmatterson committed:
Add basic extension and event support

ocdtrekkie commented:
Wow. The dream becoming reality. I remember ideas and hopes and dreams to do stuff like this many, many years ago.

Top notch, Jon!

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');
+/**
```
Can we add a space between this opening line and the docblock?

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+   *                    pass-by-reference keys to allow listener callbacks to 
+   *                    mutate
+   */
+  public function trigger($event, $eventData)
```
I think I'd like to stick a little closer to how Laravel does this for consistency.

```php
public function fire($event, $payload)
```

jonmatterson commented:
@ocdtrekkie glad you like it! Figured we might as well do this as a holdover until Nova 3!

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+    $callbackStack = [];
+    
+    // Start the pointer at the root segment.
+    $ref =& $this->callbacks;
```
Why are we using a variable named `$ref` for the callbacks? Mental weight is actually a big thing that I've been thinking about of late and not forcing someone to think as they're reading through code. When I see `$ref`, I have to stop and wonder it actually is.

Let's switch to a more descriptive variable name like `$callbacks` or `$listeners`.

jonmatterson replied:
Going to use `$callback_subarray_pointer` since it points at some subset of `$this->callbacks` depending on which iteration of the loop its within.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+      
+      // Change the pointer to the next segment.
+      $ref =& $ref['_children'][$seg];
+    };
```
One of the things I've started to do more and more with my code is to break methods down into smaller method calls that are named in a self-documenting way.

Lines 36 through 63 should probably go into a method called `_setupCallbackStack` which returns the final array and gets assigned to the `$callbackStack` variable (line 34). That means that as you read it, it's abundantly clear what's happening without needing to read comments or dig into the code.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+ * @package		Nova
+ * @category	Library
+ * @author		Anodyne Productions
+ * @copyright	2015 Anodyne Productions
```
The copyright year can be updated here

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+    $eventData['_ci_event'] = $event;
+    
+    // $callbackStack is arranged in LIFO, so array_reverse for iteration
+    foreach(array_reverse($callbackStack) as $callback)
```
Is there any reason not to reverse the array in the `_setupCallbackStack` method that I mentioned creating above?

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+   */
+  public function listen($event, $callback)
+  {  
+    $ref =& $this->callbacks;
```
This variable name should be updated to be clearer what it is.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+        }
+      }
+      
+      $seg = array_shift($event);
```
This variable name should be updated to be clearer what it is.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+  {  
+    $ref =& $this->callbacks;
+    
+    while($seg = array_shift($event))
```
This variable name should be updated to be clearer what it is.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+      if(!isset($ref['_children'][$seg]))
+      {
+        $ref['_children'][$seg] = [
+          '_children' => [],
```
What are `children` in this case?

jonmatterson replied:
The structure of `$callbacks` ends up going:

```
array(1) { 
    ["_callbacks"]=> array(0) { }
    ["_children"]=> array(1) { 
        ["db"]=> array(2) { 
            ["_callbacks"]=> array(1) { [0]=> object(Closure)#69 (2) { } } 
            ["_children"]=> array(1) { 
                ["insert"]=> array(2) { 
                    ["_callbacks"]=> array(0) { }
                    ["_children"]=> array(1) { 
                        ["prepare"]=> &array(2) { 
                                ["_children"]=> array(0) { } 
                                ["_callbacks"]=> array(1) { [0]=> object(Closure)#69 (2) { } }
                        }
                    }
                }
            }
        }
    }
}
```

Given this, when the event `['db','insert','prepare']` is called, it will first call the event registered as `['_children']['db']['_children']['insert']['_children']['prepare']['_callbacks']` and then the one for `['_children']['db']['_callbacks']`. If multiple callbacks are registered for a single event, it will call them in order for the single event.

I will add a comment for '_children' though to explain this.

agentphoenix commented on nova/modules/core/libraries/Nova_event.php:
```diff
@@ -0,0 +1,133 @@
+      {
+        $ref['_children'][$seg] = [
+          '_children' => [],
+          '_callbacks' => []
```
What are these `callbacks`? Aren't we added a callback? Is this the ability to continually nest callbacks?

jonmatterson replied:
These are the callbacks at this event depth. My comment for `_children` should explain this, but I will also add a comment for this to explain it.

agentphoenix commented on application/helpers/extension_helper.php:
```diff
@@ -0,0 +1,7 @@
+
+class Extension_container extends Nova_extension_container { }
+
+class Extension_initialization_context extends Nova_extension_initialization_context { }
```
Let's stick to one class per file.

agentphoenix replied:
No spaces between braces.

agentphoenix commented on application/helpers/extension_helper.php:
```diff
@@ -0,0 +1,7 @@
+
+require_once MODPATH.'core/helpers/Nova_extension_helper.php';
+
+class Extension_container extends Nova_extension_container { }
```
No spaces between braces.

agentphoenix commented on nova/ci/database/DB_active_rec.php:
```diff
@@ -1163,6 +1163,22 @@ public function set_insert_batch($key, $value = '', $escape = TRUE)
+			$ci->event->trigger([
+				'db',
+				'insert',
+				'prepare',
```
What are we doing with these 3 pieces of data exactly? Are there situations where we need all 3 or would omit certain ones?

jonmatterson replied:
The reason for all of these different segments is to allow someone to listen at a particular point in time. Specifically for these, you might want to do `->event->listen(['db'], ...)` to attach logging for all database calls, or you might just want to listen for `->event->listen(['db', 'insert', 'prepare'], ...)` for some specific action you take whenever you're preparing data for an insert. I didn't define another event for DB insert - at this time - but I could see us maybe wanting to add an event after the insert occurs, hence why I used `prepare` instead of it just being `['db', 'insert']`.

agentphoenix commented on nova/ci/database/DB_active_rec.php:
```diff
@@ -1163,6 +1163,22 @@ public function set_insert_batch($key, $value = '', $escape = TRUE)
 	function insert($table = '', $set = NULL)
 	{
+		$ci =& get_instance();
+		if($table && ($routerClass = $ci->router->fetch_class()) && ($routerMethod = $ci->router->fetch_method()))
```
Let's add a space between these lines for readability.

agentphoenix commented on nova/ci/database/DB_active_rec.php:
```diff
@@ -1260,6 +1276,24 @@ public function update($table = '', $set = NULL, $where = NULL, $limit = NULL)
 		$this->_merge_cache();
+		
+		$ci =& get_instance();
+		if($table && ($routerClass = $ci->router->fetch_class()) && ($routerMethod = $ci->router->fetch_method()))
```
Let's add a space between these lines for readability.

agentphoenix commented on nova/ci/database/DB_active_rec.php:
```diff
@@ -1522,6 +1556,32 @@ public function delete($table = '', $where = '', $limit = NULL, $reset_data = TR
+		$abort = false;
+		
+		$ci =& get_instance();
+		if($table && ($routerClass = $ci->router->fetch_class()) && ($routerMethod = $ci->router->fetch_method()))
```
Let's add a space between these lines for readability.

agentphoenix commented on nova/modules/core/helpers/Nova_extension_helper.php:
```diff
@@ -0,0 +1,209 @@
+		$this->_extensionContainer->required_extensions[] = $extension;
+	}
+	
+}
```
This doesn't really belong in `helpers`. Helpers are simple procedural functions that get called throughout. Since this is a full-blown class, we need to find another home for it.

jonmatterson replied:
I didn't like `libraries` for these because they're used internally by the event library and it wouldn't make sense to allow them to be included by `->load->library()` since that brings a singleton instance of the class into the context `$ci->LIBRARY_NAME`. I stuck them in helpers because they really are helpers that support the library.

agentphoenix commented on nova/modules/core/hooks/nova_utility.php:
```diff
@@ -108,4 +108,64 @@ public function maintenance()
+		foreach($extensions['enabled'] as $extensionName)
+		{
+			
+			if(!is_dir(APPPATH.'extensions/'.$extensionName.'/'))
```
Let's clean up some of these extra spaces.

agentphoenix commented on nova/modules/core/libraries/Nova_extension.php:
```diff
@@ -0,0 +1,57 @@
+<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');
+/**
```
Let's add a space above the docblock.

agentphoenix commented on nova/modules/core/libraries/Nova_extension.php:
```diff
@@ -0,0 +1,57 @@
+ * @package		Nova
+ * @category	Library
+ * @author		Anodyne Productions
+ * @copyright	2015 Anodyne Productions
```
We can update the copyright year here.

agentphoenix commented on nova/modules/core/libraries/Nova_extension.php:
```diff
@@ -0,0 +1,57 @@
+
+abstract class Nova_extension implements ArrayAccess {
+  
+  public $facades = array();
```
Let's try to stay consistent with the use of the array short syntax. We can switch this over to `[]`.

agentphoenix replied:
Why are we using the term `facades` here?

jonmatterson replied:
I was using the term `facade` because they're frontends to functionality supported by the extension. We could go with something else like `libraries` or `attached` if you'd prefer.

I will fix to using `[]`. I just noticed some places in the codebase where `array()` was used, so I just emulated that.

agentphoenix commented on nova/modules/core/libraries/Nova_location.php:
```diff
@@ -255,15 +255,29 @@ public static function view($view, $skin, $section, $data = null)
 			$location = $obj->skin.'/'.$obj->sec.'/pages/'.$obj->view;
 		}
 		
+		$ci->event->trigger(['location', 'view', 'data', $obj->sec, $obj->view], [
```
I'm actually not a big fan of this API.

I'd like to first understand why we're passing an array of values in and then figure out if there's a more eloquent way to handle something like this.

agentphoenix reviewed (changes requested):
I really appreciate your work on this and I know that working with CI can be a significant chore, so I understand how challenging this can and and appreciate your digging in to this.

This is mainly just a first pass at this stuff to ask questions and call out some things that I think could be improved. There will no doubt be further rounds of back and forth on this.

Also, I understand that some of the things I may ask for are against how things are written in Nova 2 today, but it's hard to break out of some of the better habits I've gotten in to.

jonmatterson committed:
PR feedback updates

jonmatterson commented:
@agentphoenix I have made the updates requested, with a few outstanding items:

1. I did not update the event syntax, so it's still `['db','insert','prepare','controllerclass','methodname']`. Please let me know if you want to change this to something like a dot syntax, so it would be `db.insert.prepare.controllerclass.methodname`. Internally, we'll just explode the string and still iterate over it as a set of array segments to enable to ability to listen on something like the base DB insert prepare or on a specific controller class and method name for the DB insert.
2. I did not move the `Nova_extension_helper` file. I explained in my comment reply why I do not believe it's a library in the CodeIgniter sense (it is not meant to be loaded as a singleton via `->load->library()`). Please advise if you'd like me to stick it somewhere else, but I'm unsure what would be more "CodeIgniter"-style than placing it in helpers, since those classes really are just helpers for the Event library.
3. I did not split the two classes in the `Nova_extension_helper` file into separate files. I can definitely do this, but I wanted advice on point (2) before doing that.

jonmatterson committed:
Add Template events and extension controller & CSS support

jonmatterson commented:
I just submitted another commit to this branch that:

1. Added two new event to `Nova_template::render()`, specifically `"template", "render", "data", CONTROLLER_CLASS, CONTROLLER_METHOD` and `"template", "render", "output", CONTROLLER_CLASS, CONTROLLER_METHOD`. These behave identically to their analogues in the `Nova_location` class.
2. Added support for seamless inclusion of extension-defined controllers that are created within `application/extensions/EXTENSION_NAME/controllers` and are namespace such as, for the extension `my_ext` and the controller class `my_cont`, the path will be `application/extensions/my_ext/controllers/my_cont`, the URI will be `extensions/my_ext/my_cont/METHOD_NAME/...`, and the class name will be `__extensions__my_ext__my_cont`. The reason for the underscore-separated terms is because, without namespaces, we need a convention to ensure we do not have collisions between class names in different extensions or within Nova core, and the reason for the leading underscore is explained in (3).
3. Added support for an application seamlessly overriding a class defined within an extension. To do this, using my example from (2), you add a controller to `application/controllers/extensions/my_ext/my_cont.php` with the class name `extensions__my_ext__my_cont`. This will seamlessly replace the extension-defined one with the same URI of `extensions/my_ext/my_cont/METHOD_NAME/...`. The reason for the `__` in the extension-version of this class is so that this one can extend the extension one. If they used the same class name, we could only load one or the other, and we want to be able to load both so the application one can extend the location one.
4. Added `url` method to the extension container. This supports the extension controller method URLs as defined in (2) and (3).
5. Added `inline_css` method to the extension container. This makes it possible to get CSS files defined within `application/extensions/EXTENSION_NAME/views/SEC_NAME/css`. We return them inline since we do not assume that the extension directory is web-accessible. This is useful for injecting styles for an individual page such as in [this extension](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/chronological_mission_posts).

I've also built five simple extensions that show off how some of these features can be leveraged:

* **[]()** - 
* **[]()** - 
* **[]()** - 
* **[]()** - 
https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/chronological_mission_posts


I have also added support for

jonmatterson commented:
I just submitted another commit to this branch that:

1. Added two new event to `Nova_template::render()`, specifically `"template", "render", "data", CONTROLLER_CLASS, CONTROLLER_METHOD` and `"template", "render", "output", CONTROLLER_CLASS, CONTROLLER_METHOD`. These behave identically to their analogues in the `Nova_location` class.
2. Added support for seamless inclusion of extension-defined controllers that are created within `application/extensions/EXTENSION_NAME/controllers` and are namespace such as, for the extension `my_ext` and the controller class `my_cont`, the path will be `application/extensions/my_ext/controllers/my_cont`, the URI will be `extensions/my_ext/my_cont/METHOD_NAME/...`, and the class name will be `__extensions__my_ext__my_cont`. The reason for the underscore-separated terms is because, without namespaces, we need a convention to ensure we do not have collisions between class names in different extensions or within Nova core, and the reason for the leading underscore is explained in (3).
3. Added support for an application seamlessly overriding a class defined within an extension. To do this, using my example from (2), you add a controller to `application/controllers/extensions/my_ext/my_cont.php` with the class name `extensions__my_ext__my_cont`. This will seamlessly replace the extension-defined one with the same URI of `extensions/my_ext/my_cont/METHOD_NAME/...`. The reason for the `__` in the extension-version of this class is so that this one can extend the extension one. If they used the same class name, we could only load one or the other, and we want to be able to load both so the application one can extend the location one.
4. Added `url` method to the extension container. This supports the extension controller method URLs as defined in (2) and (3).
5. Added `inline_css` method to the extension container. This makes it possible to get CSS files defined within `application/extensions/EXTENSION_NAME/views/SEC_NAME/css`. We return them inline since we do not assume that the extension directory is web-accessible. This is useful for injecting styles for an individual page such as in [this extension](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/chronological_mission_posts).

I've also built four example extensions that show off how some of these features can be leveraged (these are not final products themselves and are evolving as the underlying events and extensions framework evolves):

* **[chronological_mission_posts](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/chronological_mission_posts)** - This provides the user with the ability to pick a mission day and time (new fields) when they're creating a mission post. It also re-orders all posts show in the post lists in terms of their in-character mission day and time, as well as provides a new route `/extensions/chronological_mission_posts/sim/readposts/mission/1` that shows the mission posts in book format.
* **[bfms_character](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/bfms_character)** - This provides the user with the ability to specify a Bravo Fleet character URL, which will cause it to query bravofleet.com for character history and service record. This (1) adds new tabs to the bio view page, (2)adds a new field to the bio edit, character create and join pages, and does some other minor skeletal stuff.
* **[jquery](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/jquery)** - This provides a single class to help you generate a jQuery selector and mutator so that, as shown in `bfms_character` and `chronological_mission_posts`, you can manipulate the DOM in the `[..., "render", "output", ...]` events without having to write long JavaScript strings yourself. It does not provide any user behaviors but is a representation of what a utility extension might look like. To show off the `required_extension` functionality, this is required by both `bfms_character` and `chronological_mission_posts`.
* **[Timepicker](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions/timepicker)** - This is a really small extension that just injects CSS and JavaScript into the template so that one can designate a field as a timepicker with `data-timepicker` and encode options for the timepicker initializer into that field by JSON-encoding them in the value.

As you look through my example extensions, you'll see some potentially odd patterns, and I thought I'd explain a few of them:

* You'll note that to add content to a page, I often add my content using `$this->extensions['jquery']['generator']->select(...`. The reason for this is because Nova 2 does not provide a direct way to hook in when I need to inject additional content to some arbitrary place in the page. We could tell the user to modify their views, but the whole point is to enable authors like myself to write extensions that do not require a user to modify their base views, so this was the easiest way I could go about doing it. Basically, I just hook against a component of the view I do not expect to see a skin modify (such as `#one` tab or `.page-head`) and then load my subview into it.
* You'll note that I add inline CSS to the `javascript` region. This is because, for some reason, a `javascript` region was defined in the base skin (and thus in most other skins) but a `css` or `styles` region was not. To make these extensions as portable as possible, `javascript` region is thus the safest choice as Nova won't really work unless you include it in your skin template.

If you want to see what this enables, it makes all of these things possible without the user modifying anything except the `applications/config/extensions.php` config file:

* Added a new controller route: <img width="1081" alt="screen shot 2018-09-01 at 7 15 36 pm" src="https://user-images.githubusercontent.com/2497540/44951533-9f042880-ae1b-11e8-8fb0-71b907f32f83.png">
* Added a new button to an existing route: <img width="1035" alt="screen shot 2018-09-01 at 7 16 36 pm" src="https://user-images.githubusercontent.com/2497540/44951534-a3c8dc80-ae1b-11e8-9b73-84276d203338.png">
* Added new fields and a new formatting behavior to the mission post forms: <img width="688" alt="screen shot 2018-09-01 at 7 19 14 pm" src="https://user-images.githubusercontent.com/2497540/44951544-fc987500-ae1b-11e8-93ce-812468ccad51.png">
* Modified how the timeline field is displayed and changed how posts are ordered to be by the new mission day and time fields instead of by the post date: <img width="609" alt="screen shot 2018-09-01 at 7 22 09 pm" src="https://user-images.githubusercontent.com/2497540/44951559-64e75680-ae1c-11e8-976d-856fb4ca4cec.png">
* Added a field that fetches from an external API a service record for a character and adds a new tab for it on the character bio: <img width="1064" alt="screen shot 2018-09-01 at 7 23 26 pm" src="https://user-images.githubusercontent.com/2497540/44951565-86e0d900-ae1c-11e8-8d2d-d648498e0f69.png">

All sorts of other capabilities but just a few basic ones to show here :-)

Hope this makes some sense. I thought it would be nice to actually show a few [extensions](https://github.com/jonmatterson/nova/tree/2.5/proteus/application/extensions) in the wild as opposed to just describing or inlining them here.

jonmatterson committed:
Fix tabs and add string event name support

jonmatterson committed:
Stylistic fix

jonmatterson committed:
Override extension view bugfix and inline JS support for extensions

jonmatterson committed:
Fix template output event listener bug

jonmatterson commented:
# Events

Nova provides an Events library that allows a sim or mod to define callbacks that are run when an event occurs. While accessible throughout Nova, Events are most useful in the context of building Extensions, as Events allow Extensions to hook into existing Nova functionality.

## Event Basics

In order to define a listener, assuming `$this` is the CodeIgniter content, simply call:

```
$this->event->listen($eventName, $eventCallback);
```

The `$eventName` may be written as an array of segments or as a string with a `.` separating each. 

For example, these two `$eventName` values are identical:

* `['location', 'view', 'output', 'main', 'index']`
* `"location.view.output.main.index"`

The `$eventCallback`, meanwhile, should accept a single `$event` array parameter with data elements that will vary based on the event the listener is listening on.

Putting this all together, one could, for example, append "Hello World!" to the end of the `main/index` view as follows:

```php
$this->event->listen(['location', 'view', 'output', 'main', 'index'], function($event){
    $event['output'] .= 'Hello World!';
});
```

Or:

```php
$this->event->listen("location.view.output.main.index", function($event){
    $event['output'] .= 'Hello World!';
});
```

Further, one can define an event that will occur over an entire namespace of events.

For example, one could add controller class and method info to the body tag of the template of all pages as follows:

```php
$this->event->listen(['template', 'render', 'output'], function($event){
  $seg1 = !empty($this->uri->segment(1)) ? $this->uri->segment(1) : 'main';
  $seg2 = !empty($this->uri->segment(2)) ? $this->uri->segment(2) : 'index';
  $event['output'] = str_replace('<body', 
                                 '<body data-ci_controller="'.$seg1.'" data-ci_method="'.$seg2.'"', 
                                 $event['output']);
});
```

In the event that multiple listeners are defined, they will be executed in order of most-specific to least-specific, and if two listeners are defined for the exact same route, then the one that was attached first will run first.

For example, given these callbacks:

```php
$this->event->listen(['location', 'view', 'output'], function($event){
    $event['output'] .= 'c';
});

$this->event->listen(['location', 'view', 'output'], function($event){
    $event['output'] .= 'd';
});

$this->event->listen(['location', 'view', 'output', 'main', 'index'], function($event){
    $event['output'] .= 'a';
});
*  
$this->event->listen(['location', 'view', 'output', 'main', 'index'], function($event){
    $event['output'] .= 'b';
});
```

The following string would be added to the end of the output:

```
abcd
```

Nova provides a number of listeners listed in this article. All of these listeners support the resolution paradigm above. Each of these listeners expose different elements in their callback data, such as the data sent to a view or template, the output rendered for a view or template, the data sent to a database insert or update operation, and a flag to abort a database deletion operation.

## Available Events

#### `location.view.data.$SECTION.$VIEW`

This event allows a listener to manipulate the data that is to be sent to a view. The event array made available to the callback listener includes a `data` array, passed by reference, which can be manipulated in order to modify, append or remove data before it is passed to the view to be rendered.

Input Segments:

* `$SECTION` - The view section, such as `main` or `admin`.
* `$VIEW` - The view name, such as `personnel_character` or `characters_bio`.

Callback Data:

* `data` - The data object that will be sent to the view.
    
Example:

```php
$this->event->listen(['location', 'view', 'data', 'main', 'personnel_character'], function($event){
  $id = $this->uri->segment(3);
  $char = $this->char->get_character($id);
  $content = fms_character__api__get_character($char);
  $event['data']['fms_character'] = json_decode($content, true);
});
```

#### `location.view.output.$SECTION.$VIEW`

This event allows a listener to manipulate the output of a view once it has been rendered. The event array made available to the callback includes an `output` string, passed by reference, which can be manipulated in order to modify the output rendered by the view, or to even replace it all together. The event array also includes a `data` array of the data that was used to render the view.

Input Segments:

* `$SECTION` - The view section, such as `main` or `admin`.
* `$VIEW` - The view name, such as `personnel_character` or `characters_bio`.

Callback Data:

* `data` - The data object that was sent to the view to get the output.
* `output` - The string that was produced after `data` was sent to the view.

Example:

```php
$this->event->listen(['location', 'view', 'output', 'main', 'personnel_character'], function($event){
  $event['output'] .= $this->extension['jquery']['generator']
                  ->select('#tabs')
                  ->append($this->extension['fms_character']->view('personnel_character', $this->skin, 'main', $event['data']));
});
```

#### `template.render.data.$ROUTERCLASS.$ROUTERMETHOD`

This event allows a listener to manipulate the data that is to be sent to a template. The event array made available to the callback listener includes a `data` array, passed by reference, which can be manipulated in order to modify, append or remove data before it is passed to the template to be rendered.

Input Segments:

* `$ROUTERCLASS` - The controller class as resolved by the router.
* `$ROUTERMETHOD` - The controller method as resolved by the router.

Callback Data:

* `data` - The data object that will be sent to the view.
    
Example:

```php
$this->event->listen(['template', 'render', 'data'], function($event){
  $seg1 = !empty($this->uri->segment(1)) ? $this->uri->segment(1) : 'main';
  $seg2 = !empty($this->uri->segment(2)) ? $this->uri->segment(2) : 'index';
  $event['data']['javascript'] .= '<script type="text/javascript">
    $(document ).ready(function() { $("body").addClass("ci-route--'.$seg1.'--'.$seg2.'");});
  </script>';
});
```

#### `template.render.output.$ROUTERCLASS.$ROUTERMETHOD`

This event allows a listener to manipulate the output of a template once it has been rendered. The event array made available to the callback includes an `output` string, passed by reference, which can be manipulated in order to modify the output rendered by the template, or to even replace it all together. The event array also includes a `data` array of the data that was used to render the template.

Input Segments:

* `$ROUTERCLASS` - The controller class as resolved by the router.
* `$ROUTERMETHOD` - The controller method as resolved by the router.

Callback Data:

* `data` - The data object that was sent to the template to get the output.
* `output` - The string that was produced after `data` was sent to the template.
    
Example:

```php
$this->event->listen(['template', 'render', 'output'], function($event){
  $seg1 = !empty($this->uri->segment(1)) ? $this->uri->segment(1) : 'main';
  $seg2 = !empty($this->uri->segment(2)) ? $this->uri->segment(2) : 'index';
  $event['output'] = str_replace('<body', 
                                 '<body data-ci_controller="'.$seg1.'" data-ci_method="'.$seg2.'"', 
                                 $event['output']);
});
```

#### `db.insert.prepare.$TABLE.$ROUTERCLASS.$ROUTERMETHOD`

This event allows a listener to manipulate the data to be inserted into a database table before it is inserted. The event array made available to the callback listener includes a `data` array, passed by reference, which can be manipulated in order to modify, append or remove data before it is inserted into the table. The event array also includes a `table` string for cases where the listener is defined over the entirety of the `db.insert.prepare` namespace.

Input Segments:

* `$TABLE` - The name of the table into which the data will be written.
* `$ROUTERCLASS` - The controller class as resolved by the router.
* `$ROUTERMETHOD` - The controller method as resolved by the router.

Callback Data:

* `data` - The data to be inserted into the table.
* `table` - The name of the table.
    
Example:

```php
$this->event->listen(['db', 'insert', 'prepare', 'characters', 'main', 'join'], function($event){
  if(empty($event['data']['fms_character_url']) && $this->input->post('fms_character_url', true))
    $event['data']['fms_character_url'] = $this->input->post('fms_character_url', true);
});
```

#### `db.update.prepare.$TABLE.$ROUTERCLASS.$ROUTERMETHOD`

This event allows a listener to manipulate the data to be updated for an existing record in a database table. The event array made available to the callback listener includes a `data` array, passed by reference, which can be manipulated in order to modify, append or remove data before it is update in the table. The event array also has `where` and `limit` entries, passed by reference, which can be manipulated to adjust the scope of the update. Finally, the event array also includes a `table` string for cases where the listener is defined over the entirety of the `db.insert.prepare` namespace.

Input Segments:

* `$TABLE` - The name of the table into which the data will be written.
* `$ROUTERCLASS` - The controller class as resolved by the router.
* `$ROUTERMETHOD` - The controller method as resolved by the router.

Callback Data:

* `data` - The data to be inserted into the table.
* `table` - The name of the table.
* `where` - The where constraint defined in the update call.
* `limit` - The limit constraint defined in the update call.
    
Example:

```php
$this->event->listen(['db', 'update', 'prepare', 'missions'], function($event){
  if(empty($event['data']['mission_start']))
    $event['data']['mission_start'] = null;
  if(empty($event['data']['mission_end']))
    $event['data']['mission_end'] = null;
});
```

#### `db.delete.prepare.$TABLE.$ROUTERCLASS.$ROUTERMETHOD`

This event allows a listener to manipulate the delete operation executed on a table. The event array made available to the callback listener includes an `abort` boolean where, if set true, will prevent the delete operation from being executed. Additionally, the event array also has `where`, `limit` and `resetData` entries, passed by reference, which can be manipulated to adjust the scope of the update. Finally, the event array also includes a `table` string for cases where the listener is defined over the entirety of the `db.delete.prepare` namespace.

Input Segments:

* `$TABLE` - The name of the table into which the data will be written.
* `$ROUTERCLASS` - The controller class as resolved by the router.
* `$ROUTERMETHOD` - The controller method as resolved by the router.

Callback Data:

* `abort` - A boolean where, if set true, will prevent the delete operation from occuring.
* `table` - The name of the table.
* `where` - The where constraint defined in the delete call.
* `limit` - The limit constraint defined in the delete call.
* `resetData` - The reset_data parameter defined in the delete call.
    
Example:

```php
$this->event->listen(['db', 'delete', 'prepare', 'users'], function($event){
  if(!my_custom_auth_check()){
    $event['abort'] = true;
    redirect('my_errors/not_allowed_to_delete', 'refresh');
  }
});
```






# Extensions

Extensions are a type of Nova mod that are placed within the `application/extensions` folder and enabled in `application/config/extensions.php`. As opposed to other types of Nova mods, extensions do not require modifying any Nova code files. This simplifies installation and maintenance.

## Installing an Extension

When you download an extension, the installation will always follow roughly the same process:

1. place the extension directory within your `application/extensions` directory;
2. execute any database commands (or other actions) stipulated by the extension; and
3. add the extension to the end of `application/config/extensions.php`.

For example, if you have extracted an extension named `my_extension` into `application/extensions/my_extension`, it could be enabled by adding the following to `application/config/extensions.php`:

```php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once MODPATH.'core/config/nova_extensions.php';

$config['extensions']['enabled'][] = 'my_extension';
```

Some extensions may specify that they require another extension. If this is the case, you should ensure that all required extensions come before the extension that requires them within the `application/config/extensions.php` file.

## Building an Extension

An extension is defined as a directory placed within the `application/extensions` directory that, at minimum, has a single file `init.php`. The `init.php` file serves to bootstrap the file, attaching event listeners, defining other required extensions, etc., as necessary to achieve the extension's purpose.

As long as the extension is enabled within `application/config/extensions.php` like this:

```php
$config['extensions']['enabled'][] = 'my_extension';
```

Then the extension's `application/extensions/my_extension/init.php` will be run. All of the extension's code may be packaged directly within `init.php` or it may be included/required from other files.

### Extension `init.php`

The `init.php` file has its own scope with a number of properties and methods available within the `$this` variable.

To get the name of the extension, such as `my_extension` for `application/extensions/my_extension/init.php`:

```php
$this->name
```

To get the path of the extension directory, such as `application/extensions/my_extension` for `application/extensions/my_extension/init.php`:

```php
$this->path
```

To set another extension that must be loaded first (so specified earlier in `application/config/extensions.php`), such as requiring the `jquery` extension in another extension:

```php
$this->require_extension('jquery');
```

The two main reasons for requiring an extension:

1. an extension requires behavior enabled by another extension; or
2. an extension uses an object exposed by another extension.

For case (2), an extension can expose an object for use by other extensions as:


```php
$this->attach('generator', new jquery_generator());
```

Supposing that the above statement was within a `jquery` extension, then this exposed object can be leveraged in any extension that requires it (as well as in a sim's application code):

```php
$this->extension['jquery']['generator']->select('.page-head')->html('Hello World!')
```

Besides exposing objects, the primary thing an extension does is hook into the CodeIgniter instance to modify behavior.

To interface with the CodeIgniter instance:

```php
$this->ci
```

For example, an extension could attach an event listener such as:

```php
$this->ci->event->listen(['template', 'render', 'data'], function($event){
  $event['data']['javascript'] 
        .= '<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/timepicker/1.3.5/jquery.timepicker.min.css">'
         . '<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/timepicker/1.3.5/jquery.timepicker.min.js"></script>'
         . $this->extension['timepicker']->inline_js('timepicker', 'global');
});
```

For shorthand, calls directly to `$this` are call-forwarded to `$this->ci`.

### Extension Controllers and Views

An extension can define controllers within a `controllers` folder inside the extension folder. In order to avoid URL collisions between the extension and other routes in the Nova app, extensions will have a URL like:

```
http://my.domain.com/extensions/my_extension/my_controller/my_method
``` 

The controller would then be specified in:

```
application/extensions/my_extension/controllers/my_controller
```

To avoid a class name collision, the controller class must also be named in a special way:

* Start the class name with `__extensions`
* Next, place two more underscores `__`
* Next, place the extension name, such as `my_extension`
* Next, place two more underscores `__`
* Next, place the controller name

So for the above example, the class would be `__extensions__my_extension__my_controller`:

```php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once MODPATH.'core/libraries/Nova_controller_main.php';

class __extensions__my_extension__my_controller extends Nova_controller_main
{
	public function __construct()
	{
		parent::__construct();
		$this->_regions['nav_sub'] = Menu::build('sub', 'sim');
	}
  
  public function my_method()
  {
      $this->_regions['title'] = 'Hello World!';
      $this->_regions['content'] = 'This is an extension controller.';
      Template::assign($this->_regions);
      Template::render();
  }
}
```

To get the URL to `my_method` within the example above, one may simple call:

```php
$this->extension['chronological_mission_posts']->url('my_controller/my_method')
```

The extension makes it possible to load a view from the `views` folder within the extension folder:

```php
$this->extension['my_extension']->view('my_method_view', $this->skin, 'main', $data);
```

This view would be located at `applications/extensions/my_extension/views/main/pages/my_method_view.php`.

The extension makes it possible to load CSS from the `views` folder within the extension folder:

```php
$this->extension['my_extension']->inline_css('my_method_styles', $this->skin, 'main', $data);
```

This view would be located at `applications/extensions/my_extension/views/main/css/my_method_styles.css`, and the outcome will be returned as inline CSS within a `<style>` tag for maximum portability.

The extension makes it possible to load JavaScript from the `views` folder within the extension folder:

```php
$this->extension['my_extension']->inline_js('my_method_scripts', $this->skin, 'main', $data);
```

This view would be located at `applications/extensions/my_extension/views/main/js/my_method_scripts.js`, and the outcome will be returned as inline CSS within a `<script type="text/javascript">` tag for maximum portability.

Putting it all together, one might end up with a controller like:

```php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once MODPATH.'core/libraries/Nova_controller_main.php';

class __extensions__my_extension__my_controller extends Nova_controller_main
{
	public function __construct()
	{
		parent::__construct();
		$this->_regions['nav_sub'] = Menu::build('sub', 'sim');
	}
  
    public function my_method()
    {
        $this->_regions['title'] = 'Hello World!';
        $this->_regions['content'] = $this->extension['my_extension']->view('my_method_view', $this->skin, 'main', $data);
        $this->_regions['javascript'] .= $this->extension['my_extension']->inline_css('my_method_styles', $this->skin, 'main', $data);
        $this->_regions['javascript'] .= $this->extension['my_extension']->inline_js('my_method_scripts', $this->skin, 'main', $data);
        Template::assign($this->_regions);
        Template::render();
    }
}
```

### Overriding Extension Controllers and Views

An application can seamlessly override the behavior of an extension-defined controller without modifying the extension. To do this, suppose the goal is to override `my_controller` within `my_extension` (defined as above in `application/extensions/my_extension/controllers/my_controller.php`). To do this, define the file:

```
application/controllers/extensions/my_extension/my_controller.php
```

In this file, define the controller with the same name as the extension controller has, except without the leading `__` (the difference in class names here allows you to inherit the one from the extension). For example:

```php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once APPPATH.'extensions/my_extension/controllers/my_controller.php';

class extensions__my_extension__my_controller extends __extensions__my_extension__my_controller
{
    public function my_method()
    {
        $this->_regions['title'] = 'Hello Sky!';
        $this->_regions['content'] = $this->extension['my_extension']->view('my_method_view', $this->skin, 'main', $data);
        $this->_regions['javascript'] .= $this->extension['my_extension']->inline_css('my_method_styles', $this->skin, 'main', $data);
        $this->_regions['javascript'] .= $this->extension['my_extension']->inline_js('my_method_scripts', $this->skin, 'main', $data);
        Template::assign($this->_regions);
        Template::render();
    }
}
```

A skin can also override an extension view to provide direct support for it. For example, suppose `my_skin` wishes to provide a slightly modified version of the `my_method_view` used in the above exaple for `my_extension`. For the extension view located at , the override for the skin would be in:

```
application/views/my_skin/extensions/my_extension/main/pages/my_method_view.php
```

If a sim wishes to modify the extension view, but the skin the sim uses does not support it, an override may also be specified in:

```
application/views/_base_override/extensions/my_extension/main/pages/my_method_view.php
```
